// rollup-plugin-angular/dist/rollup-plugin-angular.js
'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var fs = _interopDefault(require('fs'));
var path = _interopDefault(require('path'));
var rollupPluginutils = require('rollup-pluginutils');

var componentRegex = /@(Component)\({([\s\S]*)}\)$/gm;
var templateUrlRegex = /templateUrl\s*:(.*)/g;
var styleUrlsRegex = /styleUrls\s*:(\s*\[[\s\S]*?\])/g;
var stringRegex = /(['"])((?:[^\\]\\\1|.)*?)\1/g;

function insertText(str, dir, preprocessor) {
  if ( preprocessor === void 0 ) preprocessor = function (res) { return res; };

  return str.replace(stringRegex, function (match, quote, url) {
    var includePath = path.join(dir, url);
    var text = fs.readFileSync(includePath).toString();
    return '`' + preprocessor(text, includePath) + '`';
  });
}

function angular (options) {
  if ( options === void 0 ) options = {};

  options.preprocessors = options.preprocessors || {};

  // ignore @angular/** modules
  options.exclude = options.exclude || [];
  if (typeof options.exclude === 'string' || options.exclude instanceof String) { options.exclude = [options.exclude]; }
  if (options.exclude.indexOf('node_modules/@angular/**') === -1) { options.exclude.push('node_modules/@angular/**'); }

  var filter = rollupPluginutils.createFilter(options.include, options.exclude);

  return {
    name: 'angular',
    transform: function transform(source, map) {
      if (!filter(map)) { return; }

      var dir = path.parse(map).dir;

      source = source.replace(componentRegex, function (match, decorator, metadata) {
        metadata = metadata
          .replace(templateUrlRegex, function (match, url) {
            return 'template:' + insertText(url, dir, options.preprocessors.template);
          })
          .replace(styleUrlsRegex, function (match, urls) {
            return 'styles:' + insertText(urls, dir, options.preprocessors.style);
          });

          return '@' + decorator + '({' + metadata + '})';
      });

      return {
        code: source,
        map: { mappings: '' }
      };
    }
  };
}

module.exports = angular;
